content/blog/apl/2021-02-28_look_at_hpkg.md

```diff
@@ -0,0 +1,119 @@
+
+```
+dir:entry : STRING : apps
+  file:type : INT : 1
+  file:atime : INT : 1551679116
+  file:mtime : INT : 1551679116
+  file:crtime : INT : 1551679116
+  file:attribute : STRING : BEOS:TYPE
+    file:attribute:type : INT : 1296649555
+    data : RAW : 31 bytes {off:0, len:31}
+  dir:entry : STRING : Tipster
+    file:permissions : INT : 493
+    file:atime : INT : 1551679116
+    file:mtime : INT : 1551679116
+    file:crtime : INT : 1551679116
+    data : RAW : 153840 bytes {off:31, len:153840}
+    file:attribute : STRING : BEOS:TYPE
+      file:attribute:type : INT : 1296649555
+      data : RAW : 35 bytes {off:153871, len:35}
+    file:attribute : STRING : BEOS:APP_SIG
+      file:attribute:type : INT : 1296649555
+      data : RAW : 26 bytes {off:153906, len:26}
+    file:attribute : STRING : BEOS:APP_FLAGS
+      file:attribute:type : INT : 1095782470
+      data : RAW : 4 bytes
+    file:attribute : STRING : BEOS:ICON
+      file:attribute:type : INT : 1447641934
+      data : RAW : 544 bytes {off:153932, len:544}
+...
+```
+
+It is possible to visualize the file structure from this debug output.  The first file entry seen here is `/apps/Tipster` and it has an HVIF formatted icon of only 544 bytes which can be obtained from the Heap at uncompressed offset 153932.  The data for the application's binary can be found in the Heap at offset 31 and the binary is 153840 bytes in length.
+
+Likewise, here is an abridged snippet of the Attributes which are fairly obvious to understand;
+
+```
+package:name : STRING : tipster
+package:summary : STRING : An application to display Haiku usage tips
+package:description : STRING : Display brief but informative ...
+package:vendor : STRING : Haiku Project
+package:packager : STRING : Builder mmlr_x86_64 <hpkg-builder@haiku-os.org>
+package:flags : INT : 0
+package:architecture : INT : 4
+package:version.major : STRING : 1
+  package:version.minor : STRING : 1
+  package:version.micro : STRING : 1
+  package:version.revision : INT : 1
+package:copyright : STRING : 2015 Vale Tolpegin
+package:copyright : STRING : 2016 Hannah Pan
+package:copyright : STRING : 2016-2018 Humdinger
+package:copyright : STRING : 2017 Akshay Agarwal
+package:copyright : STRING : 2017 vanishakesswani
+package:copyright : STRING : 2018 Janus
+package:license : STRING : MIT
+...
+```
+
+## HPKR
+
+There exists a cousin to the HPKG file format called HPKR.  The HPKR is similar but not the same as HPKG in structure.  Instead of carrying all of the files required for one single package, HPKR contains a catalog of meta-data about a number of packages for a repository such as Haiku Ports.  A repository is a collection of packages and in this way, the HPKR is the index into the repository of packages.
+
+## Finding Out More
+
+To find out more about these file formats, see the following file in the Haiku source repository;
+
+```
+/docs/develop/packages/FileFormat.rst
```

Comment:
There should be a link to https://git.haiku-os.org/haiku/tree/docs/develop/packages/FileFormat.rst here

Reply:
Done.